### ISSUE TITLE
Issues with Custom Dataset Training

### ISSUE THREAD

**Issue by theravadaz** [open-mmlab/mmaction2#1702]
Hi :) Apologies for asking what I'm sure is a very simple question, but I've been having some issues with custom dataset training for the last week, and would really appreciate some help.

I'm running my code on Google Colab Pro. I have a dataset of 2000 videos with raw frames (360 frames per video). I'm trying to train a TSN on a multi-label video activity recognition task (79 possible classes). 

As per the Google Colab tutorial, I have created a 'data' folder in my 'mmaction2' directory. In the data folder, there are two subfolders, 'train' and 'test', which contain the videos (each video is itself a folder, e.g. '00001.MP4' contains JPEG images of frames for that video). The data folder also contains a training annotation file, and a testing annotation file, e.g.:

./data/train/00001.MP4 360 4 6 27
...
...

I have imported the following configuration file: mmaction2/configs/recognition/tsn/tsn_r101_1x1x5_50e_mmit_rgb.py'. And I have made the following adjustments to it:

```
from mmcv.runner import set_random_seed

# Modify dataset type and path
cfg.dataset_type = 'RawframeDataset'
cfg.data_root = '/content/mmaction2/data/train'
cfg.data_root_val = '/content/mmaction2/data/test'
cfg.ann_file_train = '/content/mmaction2/data/train_annotation_file.txt'
cfg.ann_file_test = '/content/mmaction2/data/test_annotation_file.txt'

cfg.data.test.type = 'RawframeDataset'
cfg.data.test.ann_file = '/content/mmaction2/data/test_annotation_file.txt'
cfg.data.test.data_prefix = '/content/mmaction2/data/test/'

cfg.data.train.type = 'RawframeDataset'
cfg.data.train.ann_file = '/content/mmaction2/data/train_annotation_file.txt'
cfg.data.train.data_prefix = '/content/mmaction2/data/train/'

cfg.data.val.type = 'RawframeDataset'
cfg.data.val.ann_file = '/content/mmaction2/data/train_annotation_file.txt'
cfg.data.val.data_prefix = '/content/mmaction2/data/train/'

# The flag is used to determine whether it is omnisource training
cfg.setdefault('omnisource', False)
# Modify num classes of the model in cls_head (79 to match 79 concepts)
cfg.model.cls_head.num_classes = 79
cfg.data.train.num_classes = 79
cfg.data.val.num_classes = 79
cfg.data.test.num_classes = 79

# Set up working dir to save files and logs.
cfg.work_dir = '/content/tutorial_exps'

# The original learning rate (LR) is set for 8-GPU training.
# We divide it by 8 since we only use one GPU.
cfg.data.videos_per_gpu = cfg.data.videos_per_gpu // 16
#cfg.data.num_workers = 0
cfg.optimizer.lr = cfg.optimizer.lr / 8 / 16
cfg.total_epochs = 10

```

When I run the following code, however, I always get a bug during training of the model:

```
from mmaction.datasets import build_dataset
from mmaction.models import build_model
from mmaction.apis import train_model

import mmcv

# Build the dataset
datasets = [build_dataset(cfg.data.train)]

# Build the recognizer
model = build_model(cfg.model, train_cfg=cfg.get('train_cfg'), test_cfg=cfg.get('test_cfg'))

# Create work_dir
mmcv.mkdir_or_exist(osp.abspath(cfg.work_dir))
train_model(model, datasets, cfg, distributed=False, validate=True) <--- BUG HERE

```

The bug varies. Sometimes I'm told batch_size should be a positive integer value, but got batch_size=0. Other times, I'm told that the code was aborted because a worker was killed. Different bugs have also cropped up. I'm assuming there's something wrong with how I've loaded the data, but I can't figure out what.

I would really appreciate any help anyone is willing to give - thank you!!!

**Comment by theravadaz** [comment]
I seem to have solved the batch_size problem by changing my train, test and val pipeline to deal with a raw frame dataset rather than a video dataset. But I'm still coming up against a problem of DataLoader workers unexpectedly quitting, which causes the whole code to crash.

In PyTorch forums, people say that the solution to this problem is to set the DataLoaders num_workers to 0. How do I do this within MMAction2?

Thanks :)

**Comment by hukkai** [comment]
@theravadaz You can try to set workers_per_gpu = 0: https://github.com/open-mmlab/mmaction2/blob/12f16c15b02681a596323bafeac5b2e97cae7f3b/configs/recognition/tsn/tsn_r50_320p_1x1x3_100e_kinetics400_rgb.py#L56

**Comment by theravadaz** [comment]
> @theravadaz You can try to set workers_per_gpu = 0:
> 
> https://github.com/open-mmlab/mmaction2/blob/12f16c15b02681a596323bafeac5b2e97cae7f3b/configs/recognition/tsn/tsn_r50_320p_1x1x3_100e_kinetics400_rgb.py#L56

@hukkai Thanks so much for your reply. I've tried setting workers_per_gpu to 0, and I no longer get the 'workers quit unexpectedly' error. But my Google colab session now crashes after about 5 seconds for 'unknown reasons' whenever I start training. Do you think this is a problem with Google colab. I'm able to run the MMAction2 tutorial on colab, so I'm really confused.

Thanks once again.

**Comment by hukkai** [comment]
@theravadaz Could I have a look at the logs for 'unknown reasons'? Are you using distributed data parallel or data parallel mode?

**Comment by theravadaz** [comment]
Hi @hukkai, thanks for your reply. I'm using data parallel mode, training the model as follows:
`train_model(model, datasets, cfg, distributed=False, validate=True)`

Please find the logs here: [app.log](https://github.com/open-mmlab/mmaction2/files/9100096/app.log). The only relevant part is the last few lines: I get this warning 'src/tcmalloc.cc:283] Attempt to free invalid pointer 0x7fff1c5e8f78', and then the kernel restarts two seconds later. 

I thought it might be an issue with RAM on colab, but tried training on a much smaller dataset (20 videos) and have the same crash.

**Comment by theravadaz** [comment]
@hukkai Happy to send over the final config settings for the model, and output when model is trained (it doesn't seem to get to first epoch) if helpful :)

**Comment by hukkai** [comment]
@theravadaz Not sure what the problem is. Did you try this one [MMAction2 Tutorial.ipynb](https://colab.research.google.com/github/open-mmlab/mmaction2/blob/master/demo/mmaction2_tutorial.ipynb#scrollTo=dDBWkdDRk6oz)?

**Comment by theravadaz** [comment]
@hukkai Yep, I've tried it and it works absolutely fine. I think my best hope for debugging is making small changes to the tutorial until I hit my bug. But any suggestions you have would be very appreciated!

**Comment by theravadaz** [comment]
@hukkai Yes, it's a multi-class, multi-label classification task. I'm going to try changing the annotation files so it's just a multi-class classification task, and see what happens.

**Comment by hukkai** [comment]
@theravadaz Then you need to use BCE loss. You can refer to this: https://github.com/open-mmlab/mmaction2/blob/12f16c15b02681a596323bafeac5b2e97cae7f3b/configs/recognition/tsn/tsn_r101_1x1x5_50e_mmit_rgb.py#L19 for an example.

**Comment by hukkai** [comment]
@Dai-Wenxun @wangbo-zhao Any ideas to solve this issue? Thanks!

**Comment by theravadaz** [comment]
@hukkai @Dai-Wenxun @wangbo-zhao Hi :) I'm still getting the exact same error - when workers_per_gpu is > 0, one of my DataLoader workers is killed unexpectedly during training. If I set workers_per_gpu to 0, then the Google Colab session crashes, with uninformative logs.

However, if I run a multi-class (NOT multi-label) classification task, the training works completely fine (with workers_per_gpu = 2). This makes me think that there might be a bug in my configuration file related to the multi-label task.

However, I've been searching for bugs in the file for a week now and cannot find anything. I've attached the configuration file, and would really appreciate your help. Alternatively, is there a chance this is an issue with google colab/mmaction2?

Thank you!

```
Config:
optimizer = dict(
    type='SGD',
    constructor='TSMOptimizerConstructor',
    paramwise_cfg=dict(fc_lr5=True),
    lr=7.8125e-05,
    momentum=0.9,
    weight_decay=0.0001)
optimizer_config = dict(grad_clip=dict(max_norm=20, norm_type=2))
lr_config = dict(policy='step', step=[20, 40])
total_epochs = 10
checkpoint_config = dict(interval=5)
log_config = dict(interval=5, hooks=[dict(type='TextLoggerHook')])
dist_params = dict(backend='nccl')
log_level = 'INFO'
load_from = None
resume_from = None
workflow = [('train', 1)]
opencv_num_threads = 0
mp_start_method = 'fork'
model = dict(
    type='Recognizer2D',
    backbone=dict(
        type='ResNet',
        pretrained='torchvision://resnet101',
        depth=101,
        norm_eval=False),
    cls_head=dict(
        type='TSNHead',
        num_classes=79,
        in_channels=2048,
        spatial_type='avg',
        consensus=dict(type='AvgConsensus', dim=1),
        loss_cls=dict(type='BCELossWithLogits', loss_weight=160.0),
        dropout_ratio=0.5,
        init_std=0.01,
        multi_class=True,
        label_smooth_eps=0),
    train_cfg=None,
    test_cfg=dict(average_clips=None))
dataset_type = 'RawframeDataset'
data_root = '/content/mmaction2/data/train'
data_root_val = '/content/mmaction2/data/test'
ann_file_train = '/content/mmaction2/data/train_annotation_file.txt'
ann_file_val = '/content/mmaction2/data/test_annotation_file.txt'
ann_file_test = '/content/mmaction2/data/test_annotation_file.txt'
img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_bgr=False)
train_pipeline = [
    dict(type='SampleFrames', clip_len=1, frame_interval=1, num_clips=3),
    dict(type='RawFrameDecode'),
    dict(type='Resize', scale=(-1, 256)),
    dict(
        type='MultiScaleCrop',
        input_size=224,
        scales=(1, 0.875, 0.75, 0.66),
        random_crop=False,
        max_wh_scale_gap=1),
    dict(type='Resize', scale=(224, 224), keep_ratio=False),
    dict(type='Flip', flip_ratio=0.5),
    dict(
        type='Normalize',
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        to_bgr=False),
    dict(type='FormatShape', input_format='NCHW'),
    dict(type='Collect', keys=['imgs', 'label'], meta_keys=[]),
    dict(type='ToTensor', keys=['imgs', 'label'])
]
val_pipeline = [
    dict(
        type='SampleFrames',
        clip_len=1,
        frame_interval=1,
        num_clips=3,
        test_mode=True),
    dict(type='RawFrameDecode'),
    dict(type='Resize', scale=(-1, 256)),
    dict(type='CenterCrop', crop_size=224),
    dict(
        type='Normalize',
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        to_bgr=False),
    dict(type='FormatShape', input_format='NCHW'),
    dict(type='Collect', keys=['imgs', 'label'], meta_keys=[]),
    dict(type='ToTensor', keys=['imgs'])
]
test_pipeline = [
    dict(
        type='SampleFrames',
        clip_len=1,
        frame_interval=1,
        num_clips=25,
        test_mode=True),
    dict(type='RawFrameDecode'),
    dict(type='Resize', scale=(-1, 256)),
    dict(type='TenCrop', crop_size=224),
    dict(
        type='Normalize',
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        to_bgr=False),
    dict(type='FormatShape', input_format='NCHW'),
    dict(type='Collect', keys=['imgs', 'label'], meta_keys=[]),
    dict(type='ToTensor', keys=['imgs'])
]
data = dict(
    videos_per_gpu=2,
    workers_per_gpu=2,
    test_dataloader=dict(videos_per_gpu=1),
    train=dict(
        type='RawframeDataset',
        ann_file='/content/mmaction2/data/train_annotation_file.txt',
        data_prefix='/content/mmaction2/data/train/',
        pipeline=[
            dict(
                type='SampleFrames', clip_len=1, frame_interval=1,
                num_clips=3),
            dict(type='RawFrameDecode'),
            dict(type='Resize', scale=(-1, 256)),
            dict(
                type='MultiScaleCrop',
                input_size=224,
                scales=(1, 0.875, 0.75, 0.66),
                random_crop=False,
                max_wh_scale_gap=1),
            dict(type='Resize', scale=(224, 224), keep_ratio=False),
            dict(type='Flip', flip_ratio=0.5),
            dict(
                type='Normalize',
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_bgr=False),
            dict(type='FormatShape', input_format='NCHW'),
            dict(type='Collect', keys=['imgs', 'label'], meta_keys=[]),
            dict(type='ToTensor', keys=['imgs', 'label'])
        ],
        num_classes=79,
        multi_class=True),
    val=dict(
        type='RawframeDataset',
        ann_file='/content/mmaction2/data/test_annotation_file.txt',
        data_prefix='/content/mmaction2/data/test/',
        pipeline=[
            dict(
                type='SampleFrames',
                clip_len=1,
                frame_interval=1,
                num_clips=3,
                test_mode=True),
            dict(type='RawFrameDecode'),
            dict(type='Resize', scale=(-1, 256)),
            dict(type='CenterCrop', crop_size=224),
            dict(
                type='Normalize',
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_bgr=False),
            dict(type='FormatShape', input_format='NCHW'),
            dict(type='Collect', keys=['imgs', 'label'], meta_keys=[]),
            dict(type='ToTensor', keys=['imgs'])
        ],
        num_classes=79,
        multi_class=True),
    test=dict(
        type='RawframeDataset',
        ann_file='/content/mmaction2/data/test_annotation_file.txt',
        data_prefix='/content/mmaction2/data/test/',
        pipeline=[
            dict(
                type='SampleFrames',
                clip_len=1,
                frame_interval=1,
                num_clips=25,
                test_mode=True),
            dict(type='RawFrameDecode'),
            dict(type='Resize', scale=(-1, 256)),
            dict(type='TenCrop', crop_size=224),
            dict(
                type='Normalize',
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_bgr=False),
            dict(type='FormatShape', input_format='NCHW'),
            dict(type='Collect', keys=['imgs', 'label'], meta_keys=[]),
            dict(type='ToTensor', keys=['imgs'])
        ],
        num_classes=79,
        multi_class=True))
evaluation = dict(
    interval=5, metrics=['mmit_mean_average_precision'], save_best='auto')
work_dir = '/content/tutorial_exps'
omnisource = False
seed = 0
gpu_ids = range(0, 1)
```

**Comment by theravadaz** [comment]
@hukkai I finally solved it! I had set num_classes to 79 (indexed from 0), when there are actually 80 classes. Thanks so much for all your help with this -  I really appreciate it.  

As a sidenote, the reason I set num_classes to 79 rather than 80 is because in the dataset annotation files, total_frames is not the actual number of frames, but the number of frames indexed from 0 (e.g. my videos have 360 frames, but I had to set total_frames to 359). I assumed it would be the same for num_classes, which was the source of my confusion!

Thanks again :)
